lib/spack/spack/solver/concretize.lp

```diff
@@ -732,6 +764,39 @@ no_flags(Package, FlagType)
 #defined node_flag/3.
 #defined node_flag_set/3.
 
+
+%-----------------------------------------------------------------------------
+% Installed packages
+%-----------------------------------------------------------------------------
+% the solver is free to choose at most one installed hash for each package
+{ hash(Package, Hash) : installed_hash(Package, Hash) } 1 :- node(Package).
```

Comment:
Needs error message

Reply:
I don't think this can actually appear in any core.  If a package is in the DAG, it just says the solver has a choice of whether or not to use a hash for a package.  If it can't use a hash, it'll build.  So I think an `error()` here is a no-op.

Reply:
Imagine if two different hashes have different constraints that don't individually satisfy the requirements, but do collectively. This constraint ensures that we don't try to use both of the externals for the package at once. So I think it could be in a core for some very unusual cases.